File: Python/Search for a Range.py
# Given a sorted array of integers, find the starting and ending position of a given target value.

# Your algorithm's runtime complexity must be in the order of O(log n).

# If the target is not found in the array, return [-1, -1].

# For example,
# Given [5, 7, 7, 8, 8, 10] and target value 8,
# return [3, 4].


class Solution:
    # @param A, a list of integers
    # @param target, an integer to be searched
    # @return a list of length 2, [index1, index2]
    def searchRange(self, A, target):
		leng = len(A)
		if leng == 0:
			return [-1,-1]
		start = A[0]
		startIndex = 0
		endIndex = leng-1
		end = A[leng-1]
		while (start != end) and startIndex <=endIndex:
			if start != target and end != target:
				startIndex = startIndex + 1
				endIndex = endIndex - 1
				start = A[startIndex]
				end = A[endIndex]
				continue
			if start == target and end != target:
				endIndex = endIndex - 1
				end = A[endIndex]
			if start != target and end == target:
				startIndex = startIndex + 1
				start = A[startIndex]
		if startIndex > endIndex:
			return [-1,-1]
		if startIndex <= endIndex and start == target:
			return[startIndex,endIndex]
		if startIndex <= endIndex and start != target:
			return [-1,-1]
			

File: Java/Search_for_a_Range.java
class Search_for_a_Range{
	public int[] searchRange(int[] A, int target) {
        int[] res = {-1,-1};
        int start = 0;
        int end = A.length-1;
        while(start < end){
        	int mid = start + (end - start)/2;
        	if(A[mid] < target)
        		start = mid + 1;
        	else
        		end = mid;
        }
        int low = 0;
        if(A[start] != target)
        	return res;
        else
        	low = start;
        start = 0;
        end = A.length-1;
        while (start < end) {
            int mid = (start + end) / 2;
            if (A[mid] < target + 1)
                start = mid + 1;
            else
                end = mid;
        }
        int high = A[start] == target ? start : start - 1;
        res[0] = low;
        res[1] = high;
        return res;
    }
    public int[] searchRange(int[] A, int target) {
        int[] res = {-1,-1};
        int start = 0;
        int end = A.length-1;
        int mid = (start+end)/2;
        while(start<=end){
            mid = (start+end)/2;
            if(A[mid] == target){
                res[0]=mid;
                res[1]=mid;
                break;
            }
            if(A[mid] > target){
                end = mid - 1;
            }
            if(A[mid] < target){
                start = mid + 1;
            }
        }
        if(A[mid] != target)
            return res;
        start = mid;
        end = A.length-1;
        while(start<=end){
            int m = (start+end)/2;
            if(A[m] == target)
                start = m + 1;
            else
                end = m-1;
        }
        res[1] = end;
        start = 0;
        end = mid;
        while(start <= end){
            int m = (start+end)/2;
            if(A[m] == target)
                end = m - 1;
            else
                start = m + 1;
        }
        res[0] = start;
        return res;
    }
    public static void main(String[] agrs){
    	Search_for_a_Range s = new Search_for_a_Range();
    	int[] a = {2,2};
    	s.searchRange(a,2);
    }
}